transcom/mymove
# [MB-12887] QAE reports have the appropriate permissions model applied to them

## [Jira ticket](https://dp3.atlassian.net/browse/MB-12887) for this change
Design: https://www.figma.com/file/5AhuEkaH3bJVrtffU3wp22/Milmove-Office-App?node-id=1710%3A31160
## Summary

This PR uses permissions to prevent non-QAECSR office users from doing QAECSR stuff that they shouldnt be able to do.
- new permissions for creating, updating, and deleting evaluation reports have been applied to backend endpoints
- create report buttons have been hidden 
- if an office user tries to go to the page to edit an evaluation report (for example, by entering the URL directly), they will be instead be shown a page that explains the issue and directs them to the move details page.
- also, we fixed a big ol security hole: backend permission checks were accidentally removed in a refactoring

The permission denied page deviates a bit from the original design. We're using `<Restricted>` to conditionally render this page instead of the blocked page, rather than doing an HTTP redirect. We're also not changing the URL to `/forbidden`. It seemed like figuring out which move was being used (so that we could construct the `go to move details` button), would be hard if the move code wasn't in the URL. 

### additional scrutiny
- i think the styling is pretty close but not 100% accurate to the design. I'll take a second pass at this on Monday.
- I'm confused about what color the text on the permission denied page is supposed to be. 
- cypress tests check that a TOO is blocked from doing QAECSR stuff, is that sufficient, or should I check all the other office roles?
- are there other things that should be tested?
- CSS feedback is always appreciated

## Setup to Run Your Code

<details>
<summary>💻 You will need to use separate terminals to test this locally.</summary>

##### Terminal 1

Start the UI locally.

```sh
make client_run
```

##### Terminal 2

Start the Go server locally.

```sh
make server_run
```

</details>

### Additional steps
#### Make sure buttons are hidden
- Log in as any office user other than a QAECSR.
- Click on a move in the queue
- Go to the `Quality Assurance` tab
- Verify that you don't see any `Create report` buttons

#### Check the permission denied page
- Login as a QAECSR
- Create a report or edit an existing one, and save the URL of the page with the report form on it
  - It should look something like this: http://officelocal:3000/moves/EVLRPT/evaluation-reports/452e8013-805c-4e17-a6ff-ce90722c12c7
- Log out of the office app, and log in as a TXO or services counselor
- Visit the URL that you saved earlier, and make sure you are taken to the permission denied page.
- Verify that the `go to move details` button takes you to the move details page

#### Make sure that backend endpoints are protected 
- Get a move code for a move that has at least one shipment evaluation report (For example, `EVLRPT` in devseed data)
- Log in as a non-QAECSR office user
- Run the following code from your browser console (if you don't have devseed data, make sure you replace EVLRPT with your move code).
- Verify that you get a `401 Unauthorized` response for each of the three endpoints.
```js
function getCsrf() {
    const row = document.cookie
        .split('; ')
        .find((row) => row.startsWith('masked_gorilla_csrf='))
    const i = row.indexOf('=')
    return row.substring(i + 1)
}

// get id of first shipment evaluation report
async function getReportId(moveCode) {
    const moveResponse = await fetch(`http://officelocal:3000/ghc/v1/move/${moveCode}`)
    const move = await moveResponse.json()
    const reportsResponse = await fetch("http://officelocal:3000/ghc/v1/moves/" + move.id + "/shipment-evaluation-reports-list")
    const reports = await reportsResponse.json()
    return reports[0].id
}

async function updateReport(moveCode) {
    const reportId = await getReportId(moveCode);
    const csrf = getCsrf()
    const resp = await fetch(
        `http://officelocal:3000/ghc/v1/evaluation-reports/${reportId}`,
        {method: 'PUT', body: {}, headers: {'X-CSRF-Token': csrf, 'Content-Type': 'application/json'}}
    )
    console.log('UPDATE response status', resp.status)
    return resp
}

async function deleteReport(moveCode) {
    const reportId = await getReportId(moveCode);
    const csrf = getCsrf()
    const resp = await fetch(
        `http://officelocal:3000/ghc/v1/evaluation-reports/${reportId}`,
        {method: 'DELETE', headers: {'X-CSRF-Token': csrf}}
    )
    console.log('DELETE response status', resp.status)
    return resp
}

async function createReport(moveCode) {
    const csrf = getCsrf();
    const resp = await fetch(
        `http://officelocal:3000/ghc/v1/moves/${moveCode}/evaluation-reports`,
        {method: 'POST', headers: {'X-CSRF-Token': csrf}}
    )
    console.log('CREATE response status', resp.status)
    return resp
}

await updateReport('EVLRPT')
await deleteReport('EVLRPT')
await createReport('EVLRPT')
```
you should see something like this:
<img width="751" alt="image" src="https://user-images.githubusercontent.com/2627844/187005482-04601c42-0cf2-4c6a-809f-f8968c4ef23f.png">


## Verification Steps for Reviewers

These are to be checked by a reviewer.



### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
